File: core/src/main/javascript/entityauth/X509AuthenticationData.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>X.509 asymmetric keys entity authentication data.</p>
 *
 * <p>The X.509 certificate should be used to enumerate any entity
 * properties. The certificate subject canonical name is considered the device
 * identity. X.509 authentication data is considered equal based on the device
 * identity.</p>
 *
 * <p>
 * {@code {
 *   "#mandatory" : [ "x509certificate" ],
 *   "x509certificate" : "string"
 * }} where:
 * <ul>
 * <li>{@code x509certificate} is the Base64-encoded X.509 certificate</li>
 * </ul></p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
    "use strict";

    var EntityAuthenticationData = require('../entityauth/EntityAuthenticationData.js');
    var EntityAuthenticationScheme = require('../entityauth/EntityAuthenticationScheme.js');
    var AsyncExecutor = require("../util/AsyncExecutor.js");
    var MslEncoderException = require('../io/MslEncoderException.js');
    var MslEncodingException = require('../MslEncodingException.js');
    var MslCryptoException = require('../MslCryptoException.js');
    var MslError = require('../MslError.js');
    var X509 = require('../crypto/X509.js');
    
    var hex2b64 = require('jsrsasign').hex2b64;
    
    /**
     * Key entity X.509 certificate.
     * @const
     * @type {string}
     */
    var KEY_X509_CERT = "x509certificate";
    
    /** X.509 PEM header. */
    var PEM_HEADER = "-----BEGIN CERTIFICATE-----\n";
    /** X.509 PEM footer. */
    var PEM_FOOTER = "\n-----END CERTIFICATE-----";

    var X509AuthenticationData = module.exports = EntityAuthenticationData.extend({
        /**
         * <p>Construct a new X.509 asymmetric keys authentication data instance from
         * the provided X.509 certificate.</p>
         *
         * @param {X509} x509cert entity X.509 certificate.
         * @throws MslCryptoException if the X.509 certificate data cannot be
         *         parsed.
         */
        init: function init(x509cert) {
            init.base.call(this, EntityAuthenticationScheme.X509);

            var identity = x509cert.getSubjectString();

            // The properties.
            var props = {
                identity: { value: identity, writable: false, configurable: false },
                x509cert: { value: x509cert, writable: false, configurable: false },
            };
            Object.defineProperties(this, props);
        },

        /** @inheritDoc */
        getIdentity: function getIdentity() {
            return this.identity;
        },

        /** @inheritDoc */
        getAuthData: function getAuthData(encoder, format, callback) {
            AsyncExecutor(callback, function() {
                // Base64 encode the X.509 certificate.
                var certHex = this.x509cert.hex;
                var certB64 = hex2b64(certHex);

                // Return the authentication data.
                var mo = encoder.createObject();
                mo.put(KEY_X509_CERT, certB64);
                return mo;
            }, this);
        },

        /** @inheritDoc */
        equals: function equals(that) {
            if (this === that) return true;
            if (!(that instanceof X509AuthenticationData)) return false;
            return (equals.base.call(this, that) && this.identity == that.identity);
        },
    });

    /**
     * Construct a new RSA asymmetric keys authentication data instance from the
     * provided MSL object.
     *
     * @param {MslObject} x509AuthMo the authentication data MSL object.
     * @throws MslEncodingException if there is an error parsing the entity
     *         authentication data.
     */
    var X509AuthenticationData$parse = function X509AuthenticationData$parse(x509AuthMo) {
        var certB64;
        try {
            certB64 = x509AuthMo.getString(KEY_X509_CERT);
        } catch (e) {
            if (e instanceof MslEncoderException)
                throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "X.509 authdata" + x509AuthMo);
            throw e;
        }

        // Convert to X.509 certificate.
        var x509 = new X509();
        try {
            // Add certificate header and footer if necessary.
            certB64 = certB64.trim();
            if (!certB64.startsWith(PEM_HEADER))
                certB64 = PEM_HEADER + certB64;
            if (!certB64.endsWith(PEM_FOOTER))
                certB64 = certB64 + PEM_FOOTER;
            x509.readCertPEM(certB64);
        } catch (e) {
            throw new MslCryptoException(MslError.X509CERT_PARSE_ERROR, certB64, e);
        }
        return new X509AuthenticationData(x509);
    };
    
    // Exports.
    module.exports.parse = X509AuthenticationData$parse;
})(require, (typeof module !== 'undefined') ? module : mkmodule('X509AuthenticationData'));

File: core/src/main/java/com/netflix/msl/entityauth/X509AuthenticationData.java
/**
 * Copyright (c) 2012-2018 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.msl.entityauth;

import java.io.ByteArrayInputStream;
import java.security.cert.CertificateEncodingException;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;

import com.netflix.msl.MslCryptoException;
import com.netflix.msl.MslEncodingException;
import com.netflix.msl.MslError;
import com.netflix.msl.MslInternalException;
import com.netflix.msl.io.MslEncoderException;
import com.netflix.msl.io.MslEncoderFactory;
import com.netflix.msl.io.MslEncoderFormat;
import com.netflix.msl.io.MslObject;
import com.netflix.msl.util.Base64;

/**
 * <p>X.509 asymmetric keys entity authentication data.</p>
 * 
 * <p>The X.509 certificate should be used to enumerate any entity
 * properties. The certificate subject canonical name is considered the device
 * identity. X.509 authentication data is considered equal based on the device
 * identity.</p>
 * 
 * <p>
 * {@code {
 *   "#mandatory" : [ "x509certificate" ],
 *   "x509certificate" : "string"
 * }} where:
 * <ul>
 * <li>{@code x509certificate} is the Base64-encoded X.509 certificate</li>
 * </ul></p>
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
public class X509AuthenticationData extends EntityAuthenticationData {
    /** Key entity X.509 certificate. */
    private static final String KEY_X509_CERT = "x509certificate";
    
    /**
     * Construct a new X.509 asymmetric keys authentication data instance from
     * the provided X.509 certificate.
     * 
     * @param x509cert entity X.509 certificate.
     * @throws MslCryptoException if the X.509 certificate data cannot be
     *         parsed.
     */
    public X509AuthenticationData(final X509Certificate x509cert) throws MslCryptoException {
        super(EntityAuthenticationScheme.X509);
        this.x509cert = x509cert;
        this.identity = x509cert.getSubjectX500Principal().getName();
    }
    
    /**
     * Construct a new X.509 asymmetric keys authentication data instance from
     * the provided MSL object.
     * 
     * @param x509AuthMo the authentication data MSL object.
     * @throws MslCryptoException if the X.509 certificate data cannot be
     *         parsed.
     * @throws MslEncodingException if the X.509 certificate cannot be found.
     */
    public X509AuthenticationData(final MslObject x509AuthMo) throws MslCryptoException, MslEncodingException {
        super(EntityAuthenticationScheme.X509);
        
        // Extract X.509 certificate representation.
        final String x509;
        try {
            x509 = x509AuthMo.getString(KEY_X509_CERT);
        } catch (final MslEncoderException e) {
            throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "X.509 authdata " + x509AuthMo, e);
        }
        
        // Get the X.509 certificate factory.
        final CertificateFactory factory;
        try {
            factory = CertificateFactory.getInstance("X.509");
        } catch (final CertificateException e) {
            throw new MslInternalException("No certificate X.509 certificate factory.", e);
        }
        
        // Create X.509 cert.
        final byte[] x509bytes;
        try {
            x509bytes = Base64.decode(x509);
        } catch (final IllegalArgumentException e) {
            throw new MslCryptoException(MslError.X509CERT_INVALID, x509, e);
        }
        try {
            final ByteArrayInputStream bais = new ByteArrayInputStream(x509bytes);
            x509cert = (X509Certificate)factory.generateCertificate(bais);
            identity = x509cert.getSubjectX500Principal().getName();
        } catch (final CertificateException e) {
            throw new MslCryptoException(MslError.X509CERT_PARSE_ERROR, Base64.encode(x509bytes), e);
        }
    }
    
    /**
     * @return the X.509 certificate.
     */
    public X509Certificate getX509Cert() {
        return x509cert;
    }

    /* (non-Javadoc)
     * @see com.netflix.msl.entityauth.EntityAuthenticationData#getIdentity()
     */
    @Override
    public String getIdentity() {
        return identity;
    }
    
    @Override
    public MslObject getAuthData(final MslEncoderFactory encoder, final MslEncoderFormat format) throws MslEncoderException {
        final MslObject mo = encoder.createObject();
        try {
            mo.put(KEY_X509_CERT, Base64.encode(x509cert.getEncoded()));
        } catch (final CertificateEncodingException | IllegalArgumentException e) {
            throw new MslEncoderException("Error encoding X.509 authdata", e);
        }
        return mo;
    }
    
    /* (non-Javadoc)
     * @see com.netflix.msl.entityauth.EntityAuthenticationData#equals(java.lang.Object)
     */
    @Override
    public boolean equals(final Object obj) {
        if (obj == this) return true;
        if (!(obj instanceof X509AuthenticationData)) return false;
        final X509AuthenticationData that = (X509AuthenticationData)obj;
        return super.equals(obj) && this.identity.equals(that.identity);
    }

    /* (non-Javadoc)
     * @see com.netflix.msl.entityauth.EntityAuthenticationData#hashCode()
     */
    @Override
    public int hashCode() {
        return super.hashCode() ^ identity.hashCode();
    }

    /** Entity X.509 certificate. */
    private final X509Certificate x509cert;
    /** Entity identity. */
    private final String identity;
}
